sdk-java/src/main/java/io/littlehorse/sdk/worker/internal/LHTaskExecutor.java

```diff
@@ -0,0 +1,238 @@
+package io.littlehorse.sdk.worker.internal;
+
+import com.google.common.base.Throwables;
+import com.google.protobuf.Empty;
+import io.grpc.stub.StreamObserver;
+import io.littlehorse.sdk.common.LHLibUtil;
+import io.littlehorse.sdk.common.exception.InputVarSubstitutionError;
+import io.littlehorse.sdk.common.exception.LHSerdeError;
+import io.littlehorse.sdk.common.exception.LHTaskException;
+import io.littlehorse.sdk.common.proto.LHErrorType;
+import io.littlehorse.sdk.common.proto.LHTaskError;
+import io.littlehorse.sdk.common.proto.LittleHorseGrpc;
+import io.littlehorse.sdk.common.proto.LittleHorseGrpc.LittleHorseStub;
+import io.littlehorse.sdk.common.proto.ReportTaskRun;
+import io.littlehorse.sdk.common.proto.ScheduledTask;
+import io.littlehorse.sdk.common.proto.TaskStatus;
+import io.littlehorse.sdk.common.proto.VariableValue;
+import io.littlehorse.sdk.worker.WorkerContext;
+import io.littlehorse.sdk.worker.internal.util.VariableMapping;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Manages a queue and executor for scheduling and executing tasks in memory.
+ * Future versions will share the same queue for many workers.
+ * It handles retries with a specified delay.
+ * Internally, a {@link Semaphore} is used to prevent overwhelming the in-memory queue.
+ * Clients can control the allowed size of this queue by specifying {@code LHW_NUM_WORKER_THREADS}.
+ * Clients can also specify the number of threads dedicated to executing tasks.
+ *
+ */
+@Slf4j
+public class LHTaskExecutor {
+
+    private final ScheduledExecutorService pool;
+    private final Semaphore semaphore;
+    private static final int TOTAL_RETRIES = 5;
+
+    public LHTaskExecutor(int concurrency) {
+        this.pool = Executors.newScheduledThreadPool(concurrency);
+        this.semaphore = new Semaphore(concurrency);
+        Thread closeExecutorHook = new Thread(() -> {
+            try {
+                this.close(4, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        });
+        Runtime.getRuntime().addShutdownHook(closeExecutorHook);
+    }
+
+    /**
+     * Schedule a new task to be executed
+     */
+    public void submitTaskForExecution(
+            ScheduledTask scheduledTask,
+            LittleHorseStub specificStub,
+            List<VariableMapping> mappings,
+            Object executable,
+            Method taskMethod) {
+        try {
+            this.semaphore.acquire();
```

Comment:
This looks like we have already gotten the ScheduledTask _before_ we wait for the Semaphore. That would make us vulnerable to a lot of `TIMEOUT`s when we have a slow LHTaskMethod and a lot of tasks in the queue.